Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,170 @@
-}
 
-void __CGGradient::initWithColorComponents(const float* components, const float* locations, size_t count, CGColorSpaceRef colorspace) {
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-    DWORD componentCount = 0;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
+    inline void SetColorSpace(CGColorSpaceRef colorSpace) {
+        _colorSpace = colorSpace;
     }
 
-    _components = new float[count * componentCount];
-    memcpy(_components, components, sizeof(float) * count * componentCount);
+    inline CGColorSpaceRef GetColorSpace() {
+        return _colorSpace;
+    }
 
-    _locations = new float[count];
+    inline std::vector<CGFloat> GetColorComponents() {
+        return _colorComponents;
+    }
 
-    if (locations) {
-        memcpy(_locations, locations, sizeof(float) * count);
-    } else {
-        for (unsigned i = 0; i < count; i++) {
-            _locations[i] = (float)i / (float)(count - 1);
-        }
+    inline std::vector<CGFloat> GetStopLocations() {
+        return _stopLocations;
     }
 
-    _count = count;
-}
+    void InitWithColorComponents(const CGFloat* components, const CGFloat* locations, size_t count);
+    void InitWithColors(CFArrayRef components, const CGFloat* locations);
 
-void __CGGradient::initWithColors(CFArrayRef componentsArr, const float* locations, CGColorSpaceRef colorspace) {
-    NSArray* components = (NSArray*)componentsArr;
-    size_t componentCount = 0;
-    __CGColorSpace* cs = (__CGColorSpace*)colorspace;
-
-    _colorSpaceModel = cs->colorSpaceModel;
-
-    switch (_colorSpaceModel) {
-        case kCGColorSpaceModelRGB:
-            componentCount = 4;
-            _format = _ColorABGR;
-            break;
-        case kCGColorSpaceModelMonochrome:
-            componentCount = 2;
-            _format = _ColorGrayscale;
-            break;
-        default:
-            UNIMPLEMENTED_WITH_MSG("Unsupported colorspace used to create gradiant.");
-            break;
-    }
+private:
+    woc::StrongCF<CGColorSpaceRef> _colorSpace;
+    std::vector<CGFloat> _colorComponents;
+    std::vector<CGFloat> _stopLocations;
 
-    int count = [components count];
+    void SetLocations(const CGFloat* locations, size_t count) {
+        RETURN_IF(count == 0);
 
-    _components = new CGFloat[count * componentCount];
+        _stopLocations.reserve(count);
 
-    for (int i = 0; i < count; i++) {
-        UIColor* curColor = [components objectAtIndex:i];
+        if (locations) {
+            std::copy(locations, locations + count, std::back_inserter(_stopLocations));
```

Comment:
I don't think this is what you want.  If there are already locations in _stopLocations this will just add more rather than replacing them.  Probably want to resize _stopLocations and then just memcpy locations into _stopLocations.data()